schema/spacewalk/upgrade/susemanager-schema-4.3.0-to-susemanager-schema-4.3.1/200-rhnactionvirtcreate-refactoring.sql

```diff
@@ -0,0 +1,86 @@
+                       CASE WHEN string_agg(d.json, '') IS NOT NULL THEN CONCAT('"disks": [', string_agg(d.json, ''), '], ') END,
+                       CASE WHEN string_agg(i.json, '') IS NOT NULL THEN CONCAT('"interfaces": [', string_agg(i.json, ''), '], ') END,
+                       '"remove_disks": ', CASE WHEN a.remove_disks = 'Y' THEN 'true' ELSE 'false' END, ', ',
+                       '"remove_interfaces": ', CASE WHEN a.remove_interfaces = 'Y' THEN 'true' ELSE 'false' END,
+                     '}'
+                   ) AS json
+                FROM rhnactionvirtcreate a
+                LEFT JOIN (
+                  SELECT action_id, CONCAT(
+                    '{',
+                        CASE WHEN bus IS NOT NULL THEN CONCAT('"bus": "', bus, '", ') END,
+                        CASE WHEN format IS NOT NULL THEN CONCAT('"format": "', format, '", ') END,
+                        CASE WHEN pool IS NOT NULL THEN CONCAT('"pool": "', pool, '", ') END,
+                        CASE WHEN source_file IS NOT NULL THEN CONCAT('"source_file": "', source_file, '",') END,
+                        '"size": ', size, ', '
+                        '"device": "', device, '"'
+                    '}'
+                  ) AS json
+                  FROM rhnactionvirtcreatediskdetails
+                ) AS d ON a.action_id = d.action_id
+                LEFT JOIN (
+                  SELECT action_id, CONCAT(
+                    '{',
+                      CASE WHEN mac IS NOT NULL THEN CONCAT('"mac": "', mac, '", ') END,
+                      '"type": "', type, '", '
+                      '"source": "', source, '"'
+                    '}'
+                  ) AS json
+                  FROM rhnactionvirtcreateinterfacedetails
+                ) AS i ON a.action_id = i.action_id
+                GROUP BY a.action_id
+            ) AS action_details
+            WHERE action_id = action_details.id;
+            
+            -- Remove the now useless columns and tables
+            ALTER TABLE rhnactionvirtcreate
+                DROP COLUMN IF EXISTS uuid CASCADE,
+                DROP COLUMN IF EXISTS vm_type CASCADE,
+                DROP COLUMN IF EXISTS vm_name CASCADE,
+                DROP COLUMN IF EXISTS os_type CASCADE,
+                DROP COLUMN IF EXISTS memory,
+                DROP COLUMN IF EXISTS vcpus,
+                DROP COLUMN IF EXISTS arch CASCADE,
+                DROP COLUMN IF EXISTS graphics_type CASCADE,
+                DROP COLUMN IF EXISTS remove_disks,
+                DROP COLUMN IF EXISTS remove_interfaces,
+                DROP COLUMN IF EXISTS cobbler_system CASCADE,
+                DROP COLUMN IF EXISTS kickstart_host CASCADE,
+                DROP COLUMN IF EXISTS kernel_options CASCADE,
+                DROP COLUMN IF EXISTS cluster_definitions;
+            DROP TABLE IF EXISTS rhnactionvirtcreatediskdetails, rhnactionvirtcreateinterfacedetails;
```

Comment:
seems these tables have sequences and indexes. Check if they need to be dropped as well.

Reply:
Right. I usually forget about them.